### PR TITLE
Backup and Restore

### PR DESCRIPTION
**Pull request by DeeAjayi**
This PR:

- adds information about Backing Up and Restoring Postgres DB created by Harbor

### CODE REVIEW

File: CLONE_RESTORE.md [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    uid: ""
+    cluster: ""
+```
+We have to find a way to simulate what this parameter does by passing in the needed parameters in the `pod-config` configmap. These are the parameters needed to trigger the clone scripts 
+on postgres pod startup
+
+
+PS: This is the ideal solution for the restoring a base backup as they were tested with regular postgres clusters.
+
+In other to restore backups from s3 these are the steps to follow:
```
Comment (MarcelMue) [inline CLONE_RESTORE.md:134]:
```suggestion
In order to restore backups from s3 these are the steps to follow:
```

File: CLONE_RESTORE.md [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+5. Create the Harbor CR
+   1. Create a new namespace to deploy harbor in 
+      `kubectl create namespace harbor-cluster`
+   2. Create a txt file that holds your AWS secretkey and create a kubernetes secret using the file
+      `kubectl create secret -n harbor-cluster generic secretkey --from-file=secret=./password.txt`
+   3. You can use this reference manifest file to configure the harbor cluster and it's dependencies [here](https://docs.google.com/document/d/1g1sKP0fG-2qC9aLRfdMv75zJZ_B-S3I1B3upeBKgu0A)
+   4. At this point, the pods will fail, reference problem 2 under the "Challenges faced while implementing this solution with Harbor Operator" to resolve it.
+   5. Delete the `harbor-operator-postgres-operator-*` pod in the `harbor-operator-ns` namespace and wait for the pods to be recreated.
+
+To confirm if backups are being created by run `envdir /run/etc/wal-e.d/env wal-g backup-list` in the postgres pod created, you'll see a list of backups in s3
```
[the remark this answers is not in the document]
Comment (DeeAjayi) [reply CLONE_RESTORE.md:116]: Yes, that is correct. Updated the README to reflect it

File: CLONE_RESTORE.md [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+   CLONE_USE_WALG_RESTORE: "true"
+   USEWALG_RESTORE: "true"
+   CLONE_AWS_REGION: ""
+   CLONE_AWS_ACCESS_KEY_ID:
+   CLONE_AWS_SECRET_ACCESS_KEY:
+   ```
+3. You can use this reference manifest file to configure the harbor cluster and it's dependencies [here](https://docs.google.com/document/d/1g1sKP0fG-2qC9aLRfdMv75zJZ_B-S3I1B3upeBKgu0A)
+4. At this point, the pods will fail, reference problem 2 under the "Challenges faced while implementing this solution with Harbor Operator" to resolve it.
+5. Delete the `harbor-operator-postgres-operator-*` pod in the `harbor-operator-ns` namespace and wait for the pods to be recreated.
+6. Go to the configured harbor url (set as externalURL in the HarborCluster CR) to confirm your old data has been restored
```
[the remark this answers is not in the document]
Comment (DeeAjayi) [reply CLONE_RESTORE.md:156]: Yes but a bit tricky.
- We can update the value of the CLONE parameters to that of the current running cluster so that it can do DR automatically on subsequent times the pods go down (This can be the default) the only thing to worry about is if the entire HarborCluser CR is deleted then we will need to update the `CLONE_WAL_BUCKET_SCOPE_SUFFIX` parameter to the old cluster ID and change it back to the current one when a successful backup has been done.

File: CLONE_RESTORE.md [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+USE_WALG_BACKUP: "true"
+USE_WALG_RESTORE: "true"
+BACKUP_NUM_TO_RETAIN: "14" # For 2 backups per day, keep 7 days of base backups
+BACKUP_SCHEDULE: "*/10 * * * *"
+AWS_ACCESS_KEY_ID: < * >
+AWS_SECRET_ACCESS_KEY: < * >
+AWS_REGION: < * >
+WALG_DISABLE_S3_SSE: "true"
+WALE_DISABLE_S3_SSE: "true"
+WAL_S3_BUCKET: < * > # Required, the bucket the backups will be created in
```
Comment (cagiti) [inline CLONE_RESTORE.md:25]: Have the following `harbor-operator` logical_backup properties been tested as alternative method for configuring backups to s3?

File: CLONE_RESTORE.md [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+   CLONE_HOST: "" # The same value as CLONE_SCOPE
+   CLONE_USER: "postgres"
+   CLONE_WAL_S3_BUCKET: "" # s3 bucket name the backup is stored in
+   CLONE_WAL_BUCKET_SCOPE_SUFFIX: "" # This is the id of the cluster, usually in the clone_scope folder in s3 (s3://<bucket-name>/spilo/<clone-scope>/<scope-suffix>) and it has to start with "/" e.g "/99e85b8a-3f92-4c86-8d42-5056348c1e7e"
+   CLONE_PORT: "5432"
+   CLONE_USE_WALG_RESTORE: "true"
+   USEWALG_RESTORE: "true"
+   CLONE_AWS_REGION: ""
+   CLONE_AWS_ACCESS_KEY_ID:
+   CLONE_AWS_SECRET_ACCESS_KEY:
```
Comment (cagiti) [inline CLONE_RESTORE.md:156]: It doesn't look like we can set this using an alternative method right now, but does seem like we could make some changes upstream to do so as the postgres-operator accepts [clone parameters](https://github.com/zalando/postgres-operator/blob/920f3dee3e4fa0fdf244a12312a219061ed38ba8/charts/postgres-operator/crds/postgresqls.yaml#L110-L135), which could be configured in the [postgres-spec](https://github.com/goharbor/harbor-operator/blob/master/pkg/cluster/controllers/database/generate.go#L54-L71)

File: CLONE_RESTORE.md [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+<h2> Creating and storing backups in S3 using WALG </h2> 
+In other to create backups stored in s3 the following fields must be set in the `pod-config` configmap the postgres operator will read environment variables from.
+
+```
+USE_WALG_BACKUP: "true"
+USE_WALG_RESTORE: "true"
+BACKUP_NUM_TO_RETAIN: "14" # For 2 backups per day, keep 7 days of base backups
+BACKUP_SCHEDULE: "*/10 * * * *"
+AWS_ACCESS_KEY_ID: < * >
+AWS_SECRET_ACCESS_KEY: < * >
```
Comment (cagiti) [inline CLONE_RESTORE.md:21]: perhaps we should be setting a [secret](https://github.com/giantswarm/harbor-operator-app/blob/a1333bd7599ddfa8f5c7fb6ce26dc777b84fa7b3/helm/harbor-operator/charts/postgres-operator/values.yaml#L126) rather than a configmap for these values

File: CLONE_RESTORE.md (hunk 4 above)
Comment (DeeAjayi) [reply CLONE_RESTORE.md:25]: Yeah, we looked at it but it doesn't meet our [use-case](https://opensource.zalando.com/postgres-operator/docs/administrator.html#logical-backups). Also, it requires the `enableLogicalBackup` parameter to be set to true which isn't currently in Harbor operator

**Comment by MarcelMue** [comment]
@piontec Please remember to take a look here.

**Review by kubasobon — COMMENTED** [review]
Current version reads OK to me. I'll wait to discuss with @piontec after he gets through it.

File: CLONE_RESTORE.md [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+   CLONE_WAL_S3_BUCKET: "" # s3 bucket name the backup is stored in
+   CLONE_WAL_BUCKET_SCOPE_SUFFIX: "" # This is the id of the cluster, usually in the clone_scope folder in s3 (s3://<bucket-name>/spilo/<clone-scope>/<scope-suffix>) and it has to start with "/" e.g "/99e85b8a-3f92-4c86-8d42-5056348c1e7e"
+   CLONE_PORT: "5432"
+   CLONE_USE_WALG_RESTORE: "true"
+   USEWALG_RESTORE: "true"
+   CLONE_AWS_REGION: ""
+   CLONE_AWS_ACCESS_KEY_ID:
+   CLONE_AWS_SECRET_ACCESS_KEY:
+   ```
+3. You can use this [reference manifest file](https://docs.google.com/document/d/1g1sKP0fG-2qC9aLRfdMv75zJZ_B-S3I1B3upeBKgu0A) to configure the harbor cluster and its dependencies.
```
Comment (piontec) [inline CLONE_RESTORE.md:198]: Again, please include all the docs in a single place

File: CLONE_RESTORE.md [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+1. **Problem:** The `s3` path wasn't correct, majorly because the tools within the postgres pod (WALE, WALG, PATRONI) were outdated.
+
+   **Solution:** Update the spilo image version used by the postgres operator to a more recent version and that should fix the issue.
+
+
+2. **Problem:** `FATAL:  no pg_hba.conf entry for host "[local]", user "postgres", database "postgres", SSL off`. This was caused because the Harbor Operator hardcoded the [pg_hba.config](https://github.com/goharbor/harbor-operator/blob/master/pkg/cluster/controllers/database/generate.go#L94) and it doesn't include the `postgres` user access to make replications.
+
+   **Solution:** `This is a possible solution that needs to be pushed upstream`.
+
+    To resolve this, edit the postgres operator CR and included the following under the patroni:pg_hba section and delete the pods.
```
Comment (piontec) [inline CLONE_RESTORE.md:243]:
```suggestion
    To resolve this, edit the postgres operator CR and include the following under the patroni:pg_hba section and delete the pods.
```

File: CLONE_RESTORE.md [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+    ```
+
+3. **Problem:** The `clone_with_wale.py` script doesn't use the correct `s3` path to download the backups for restoration.
+
+   **Solution:** The `CLONE_AWS_ENDPOINT` environment variable affects the `s3` path so remove it if you have it.
+
+
+### Post Recovery Options
+
+After a database has been recovered successfully from a backup from s3, the CLONE parameters can be removed from the `pod-config` configmap or left alone, this is because the postgresql service will always pick the latest data available regardless if the CLONE parameters are specified or not.
```
Comment (piontec) [inline CLONE_RESTORE.md:267]: Sorry, I don't get this point. What will or won't work and how?

File: CLONE_RESTORE.md [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+
+## Summary
+
+There exists a solution for backup/restore of postgres. The solution is slightly less awkward for deployment solutions utilising the [zalando postgres operator](https://github.com/zalando/postgres-operator) directly, however it involves more operational overheard since the database is then decoupled from Harbor. The _issues_ outlined above could be resolved to remove friction of using the internal postgres operator (_which still is the zalando postgres operator, but it's configuration isn't as easily configurable_).
+
+Some important notes to consider:
+- Backups stored within the Postgres Backup S3 Bucket should be verified with backups listed by the `wal-g` cli within a Production environment setting (_alerting when the condition fails_).
```
Comment (piontec) [inline CLONE_RESTORE.md:274]: How to do that?

File: CLONE_RESTORE.md [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+- The standby `postgres` secret (_along with other kubernetes secrets_) should be replicated or backed up outside of the cluster.
```
Comment (piontec) [inline CLONE_RESTORE.md:275]: Why? How important is that?

**Review by piontec — COMMENTED** [review]
In general: can we split this into separate 'deployment' and 'recovery' docs? THey are a bit entangled here and as a "recovery under a stress" doc it might be hard to follow and filter out what is needed

File: CLONE_RESTORE.md [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+       redis-operator:
+         enabled: true
+       minio-operator:
+         enabled: false
+       postgres-operator:
+         enabled: true
+         image:
+           tag: v1.8.2
+         configGeneral:
+           docker_image: registry.opensource.zalan.do/acid/spilo-14:2.1-p7
```
[the remark this answers is not in the document]
Comment (cagiti) [reply CLONE_RESTORE.md:56]: We are in the process of updating upstream, we have to ensure a migration path exists for existing users. We have an existing [PR](https://github.com/goharbor/harbor-operator/pull/980) for this.

File: CLONE_RESTORE.md [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+To be able to recover harbor indexed data in case of disaster so we can still have the list of cached artifacts stored in harbor
+
+There are currently two ways to configure a database in harbor
+1. Standard database configurations can be used to set the existing pre-deployed or cloud database services as the dependent database of the deploying Harbor.
+1. The in-cluster database configuration can be configured to let the **Harbor operator automatically deploy an in-cluster PostgreSQL database service using the zalando postgres operator** with HA supported as the dependent database of the deploying Harbor.
+
+The second method is the preferred method because it removes the need to manage a separate postgres instance.
+
+# Creating and storing backups in S3 using WALG
+In other to create backups stored in s3 the following fields must be set in the `pod-config` configmap the postgres operator will read environment variables from.
```
[the remark this answers is not in the document]
Comment (cagiti) [reply CLONE_RESTORE.md:14]: thanks

File: CLONE_RESTORE.md (hunk 9 above)
Comment (cagiti) [reply CLONE_RESTORE.md:267]: I don't think this is required.

File: CLONE_RESTORE.md (hunk 10 above)
Comment (cagiti) [reply CLONE_RESTORE.md:274]: I'd suggest the creation of an backup/restore system test when a repository containing the values associated with the `harbor-operator-app` and configuration values for `harbor-cluster` are known.

File: CLONE_RESTORE.md (hunk 11 above)
Comment (cagiti) [reply CLONE_RESTORE.md:275]: It's important to note where the stateful data is stored, this helps an operator understand where the important data associated with the backup is located.

Comment (piontec) [reply CLONE_RESTORE.md]: Sure, but I still don't get what should I do here? This seems an important step, but how do I execute it?

**Review by MarcelMue — APPROVED** [review]
I think we should get this merged and iterate.
